torch/nn/modules/lazy.py

```diff
@@ -0,0 +1,116 @@
+import itertools
+
+import torch
+from ..parameter import UninitializedParameter
+
+
+class LazyModuleMixin:
+    """A mixin for modules that lazily initialize buffers and parameters.
+
+    Unlike regular modules, subclasses of this module can initialize
+    buffers and parameters outside of the constructor (``__init__``).
+    This allows you to, for example, initialize parameters in ``forward``
+    method to determine the shape of the weight based on the initial input.
+    Be sure to run "dummy" forward once to initialize all parameters in case
+    that you need to access the weights before the actual real ``forward`` pass.
+    Note that lazy modules cannot validate if the shape is correct during
+    deserialization.  Also note that the initial weights values may become different
+    from the original (non-lazy) module even if the random seed is manually
+    configured, as the order of initialization is different from the original
+    one; especially, ``module.cuda()`` may cause the initialization to run on
+    a GPU.
+    The default value of lazy buffers and parameters are ``torch.Tensor([])``
+    and ``UninitializedParameter()``, respectively.
```

Comment:
I feel like this is missing examples of how one would use LazyModule in common scenarios, e.g. what if one wants to do custom initialization? DataParallel? Reparameterization such as weight_norm or spectral_norm? Common case where none of the above is needed?

Reply:
`DataParallel` and `weight_norm` or `spectral_norm` do not work without a dummy pass first, or unless we change their code to deal with uninitialized parameters.  I think we can detect these situations easily and just raise an error.
asking the user to do a dummy forward on everything that tries to access the uninitialized parameter data.

Reply:
It's ok that they don't work, the ask here is for documentation and examples (as @albanD and @yf225 mentioned)